crates/viewer/re_renderer/src/video/player.rs

```diff
@@ -505,42 +545,69 @@ impl VideoPlayer {
-    requested_sample: Option<&re_video::SampleMetadata>,
-    last_decoded_frame_pts: Time,
-) -> DecoderDelayState {
-    let up_to_date =
-        requested_sample.is_some_and(|s| s.presentation_timestamp == last_decoded_frame_pts);
-
-    match previous_decoder_delay_state {
-        DecoderDelayState::UpToDate | DecoderDelayState::UpToDateWithinTolerance => {
-            if up_to_date {
-                DecoderDelayState::UpToDate
-            } else if is_significantly_behind(
-                video_description,
-                requested_sample,
-                last_decoded_frame_pts,
+    /// Given the current decoder delay state, update it based on the new requested frame and the last decoded frame.
+    #[must_use]
+    fn determine_new_decoder_delay_state(
+        &self,
+        video_description: &re_video::VideoDataDescription,
+        requested_sample: Option<&re_video::SampleMetadata>,
+        last_decoded_frame_pts: Time,
+    ) -> DecoderDelayState {
+        let Some(requested_sample) = requested_sample else {
+            // Desired sample doesn't exist. This should only happen if the video is being GC'ed from the back.
+            // We're technically not catching up, but we may as well behave as if we are.
+            return DecoderDelayState::Behind;
+        };
+
+        if requested_sample.presentation_timestamp == last_decoded_frame_pts {
+            return DecoderDelayState::UpToDate;
+        }
+
+        // If we're streaming in live video, we're a bit more relaxed about what counts as "catching up" for newly incoming frames:
+        // * we don't want to show the spinner too eagerly and rather give the impression of a delayed stream
+        // * some decoders need a certain amount of samples in the queue to produce a frame.
+        //   See AsyncDecoder::min_num_samples_to_enqueue_ahead for more details about decoder pecularities.
+        let recently_updated_video = video_description
+            .last_time_updated_samples
+            .is_some_and(|t| t.elapsed() < TIME_UNTIL_VIDEO_ASSUMED_ENDED);
+        if recently_updated_video {
+            let min_num_samples_to_enqueue_ahead =
+                self.sample_decoder.min_num_samples_to_enqueue_ahead();
+            let allowed_delay =
+                min_num_samples_to_enqueue_ahead.max(TOLERATED_OUTPUT_DELAY_IN_NUM_FRAMES);
+
+            let sample_idx_end = video_description.samples.next_index();
+            for (_, sample) in video_description.samples.iter_index_range_clamped(
+                &(sample_idx_end.saturating_sub(allowed_delay)..sample_idx_end),
             ) {
-                DecoderDelayState::Behind
-            } else {
-                DecoderDelayState::UpToDateWithinTolerance
+                if sample.presentation_timestamp == last_decoded_frame_pts {
+                    return DecoderDelayState::UpToDateToleratedEdgeOfLiveStream;
+                }
             }
         }
```

Comment:
this block is new, everything else in this method is just moving around stuff and shouldn't affect behavior
(I did in fact try a bit with this block commented out to check if that's sane as expected)